kedro/extras/datasets/pandas/generic_dataset.py

```diff
@@ -0,0 +1,242 @@
+"""``GenericDataSet`` loads/saves data from/to a data file using an underlying
+filesystem (e.g.: local, S3, GCS). It uses pandas to handle the
+type of read/write target.
+"""
+from copy import deepcopy
+from pathlib import PurePosixPath
+from typing import Any, Dict
+
+import fsspec
+import pandas as pd
+
+from kedro.io.core import (
+    AbstractVersionedDataSet,
+    DataSetError,
+    Version,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+NON_FILE_SYSTEM_TARGETS = [
+    "clipboard",
+    "numpy",
+    "sql",
+    "period",
+    "records",
+    "timestamp",
+    "xarray",
+    "sql_table",
+]
+
+
+class GenericDataSet(AbstractVersionedDataSet):
+    """`pandas.GenericDataSet` loads/saves data from/to a data file using an underlying
+    filesystem (e.g.: local, S3, GCS). It uses pandas to dynamically select the
+    appropriate type of read/write target on a best effort basis.
+
+    Example using `YAML API
+    <https://kedro.readthedocs.io/en/stable/05_data/\
+        01_data_catalog.html#using-the-data-catalog-with-the-yaml-api>`_:
+
+    .. code-block:: yaml
+
+        >>> cars:
+        >>>   type: pandas.GenericDataSet
+        >>>   file_format: csv
+        >>>   filepath: s3://data/01_raw/company/cars.csv
+        >>>   load_args:
+        >>>     sep: ","
+        >>>     na_values: ["#NA", NA]
+        >>>   save_args:
+        >>>     index: False
+        >>>     date_format: "%Y-%m-%d"
+
+    This second example is able to load a SAS7BDAT file via the :code:`pd.read_sas` method.
+    Trying to save this dataset will raise a `DataSetError` since pandas does not provide an
+    equivalent :code:`pd.DataFrame.to_sas` write method.
+
+    .. code-block:: yaml
+
+        >>> flights:
+        >>>    type: pandas.GenericDataSet
+        >>>    file_format: sas
+        >>>    filepath: data/01_raw/airplanes.sas7bdat
+        >>>    load_args:
+        >>>       format: sas7bdat
+
+    Example using Python API:
+    ::
+
+        >>> from kedro.extras.datasets.pandas import GenericDataSet
+        >>> import pandas as pd
+        >>>
+        >>> data = pd.DataFrame({'col1': [1, 2], 'col2': [4, 5],
+        >>>                      'col3': [5, 6]})
+        >>>
+        >>> # data_set = GenericDataSet(file_format='csv', filepath="s3://test.csv")
+        >>> data_set = GenericDataSet(file_format='csv', filepath="test.csv")
+        >>> data_set.save(data)
+        >>> reloaded = data_set.load()
+        >>> assert data.equals(reloaded)
+
+    """
+
+    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
+    DEFAULT_SAVE_ARGS = {}  # type: Dict[str, Any]
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        file_format: str,
+        filepath: str,
```

Comment:
Nit: Maybe `filepath` can be the first argument, to be consistent with the other datasets.